libs/common/Networking/TcpNetworkHandlerBase.cs

```diff
@@ -135,31 +136,48 @@ void Dispose(SocketAsyncEventArgs e)
         }
 
         void RecvEventArg_Completed(object sender, SocketAsyncEventArgs e)
+        {
+            if (e.BytesTransferred == 0 || e.SocketError != SocketError.Success || serverHook.Disposed)
+            {
+                // No more things to receive
+                Dispose(e);
+                return;
+            }
+            // Complete receive event and release thread while we process data async
+            _ = HandleReceiveAsync(sender, e);
+        }
+
+        private async ValueTask HandleReceiveAsync(object sender, SocketAsyncEventArgs e)
         {
             try
             {
-                do
-                {
-                    if (e.BytesTransferred == 0 || e.SocketError != SocketError.Success || serverHook.Disposed)
-                    {
-                        // No more things to receive
-                        Dispose(e);
-                        break;
-                    }
-                    OnNetworkReceive(e.BytesTransferred);
-                    e.SetBuffer(networkReceiveBuffer, networkBytesRead, networkReceiveBuffer.Length - networkBytesRead);
-                } while (!e.AcceptSocket.ReceiveAsync(e));
+                await OnNetworkReceiveAsync(e.BytesTransferred);
```

Comment:
Call await only if valuetask did not finish synchronously, in order to avoid await overhead in sync paths.